src/Geno/C++/GUI/Widgets/TextEdit.h

```diff
@@ -174,6 +176,151 @@ class TextEdit
 		}
 	};
 
+	enum class CommandType
+	{
+		None,
+		EnterText,
+		Backspace,
+		Del,
+		DeleteLines,
+		InsertLines,
+		Enter,
+		Tab,
+		TabAllCursors,
+		RestoreCursor,
+		Delete,
+		SwapLines
+	};
+
+	class SubCommand
+	{
+	public:
+		SubCommand( CommandType Type )
+			: Type( Type )
+			, CursorIndex( -1 )
+			, Character( 0, 0 )
+			, NoMove( false )
+		{
+		}
+
+		CommandType Type;
+
+		Cursor CursorCopy;
+		int    CursorIndex;
+
+		Coordinate Start = Coordinate( -1, -1 );
+		Coordinate End   = Coordinate( -1, -1 );
+
+		std::vector< Line > Lines;
+		Glyph               Character;
+
+		union
+		{
+			bool NoMove;
+			bool Shift;
+			bool IgnoreSelection;
+			bool Up;
+		};
+	};
+
+	class Command
+	{
+	public:
+		CommandType Type;
+
+		Command( CommandType Type )
+			: Type( Type )
+		{
+		}
+
+		~Command()
+		{
+			ClearSubCommands();
+		}
+
+		void AddSubCommand( SubCommand* pSubCommand )
+		{
+			SubCommands.push_back( pSubCommand );
+		}
+
+		void ClearSubCommands()
+		{
+			for( SubCommand* pSub : SubCommands )
+			{
+				delete pSub;
+			}
+
+			SubCommands.clear();
+		}
+
+		std::vector< Cursor >      Cursors;
+		std::vector< SubCommand* > SubCommands;
+		BoxModeDirection           BoxModeDir;
+		CursorInputMode            CursorMode;
+		MultiCursorMode            CursorMultiMode;
+	};
+
+	struct File;
+
+	class CommandStack
+	{
+	public:
+		Command* GetCommand( File* pFile, CommandType Type = CommandType::None )
+		{
+			if( pCurrentCommand ) return pCurrentCommand;
+
+			pCurrentCommand = new Command( Type );
+
+			pCurrentCommand->BoxModeDir      = pFile->BoxModeDir;
+			pCurrentCommand->CursorMode      = pFile->CursorMode;
+			pCurrentCommand->CursorMultiMode = pFile->CursorMultiMode;
+
+			return pCurrentCommand;
+		}
+
+		void FinishCommand( const std::vector< Cursor >& rCursors )
+		{
+			GENO_ASSERT( pCurrentCommand != nullptr );
+
+			if( pCurrentCommand->SubCommands.size() == 0 )
+			{
+				delete pCurrentCommand;
+				pCurrentCommand = nullptr;
+				return;
+			}
+
+			int Count = ( int )Commands.size();
+
+			if( CommandInProcess )
+			{
+				Commands[ Location ] = pCurrentCommand;
+			}
+			else if( Location == Count - 1 )
+			{
+				Commands.push_back( pCurrentCommand );
+				Location++;
+			}
+			else if( Location < ( int )Count )
+			{
+				for( int i = Location + 1; i < ( int )Commands.size(); i++ )
+					delete Commands[ i ];
+
+				Commands.erase( Commands.begin() + ( Location + 1 ), Commands.end() );
+				Commands.push_back( pCurrentCommand );
+				Location++;
+			}
+
+			pCurrentCommand->Cursors = rCursors;
+			pCurrentCommand          = nullptr;
+		}
+
+	public:
+		bool                    CommandInProcess = false;
+		Command*                pCurrentCommand  = nullptr;
+		int                     Location         = -1;
+		std::vector< Command* > Commands;
+	};
```

Comment:
This is a lot of code in the header file. Do these classes even need to be declared in the header-file at all? Can't they be relayed to the .cpp file only?

Reply:
I'm planning on fixing the entire header file so I just dumped it there for now. It's not gonna stay like that later.